cms/topics/utils.py

```diff
@@ -26,3 +54,256 @@ def format_time_series_as_document_list(time_series: StreamValue) -> list[dict[s
+        descendant_series = ArticleSeriesPage.objects.descendant_of(self.topic_page)
+        latest_pks = self._get_latest_articles_by_series(descendant_series)
+
+        # Get the latest articles under this topic page (topic page -> articles index -> article series page ->
+        # latest stats article page), excluding those already highlighted
+        return list(
+            StatisticalArticlePage.objects.filter(pk__in=latest_pks)
+            .exclude(pk__in=excluded_pks)
+            .live()
+            .public()
+            .defer_streamfields()
+            .order_by("-release_date")
+        )
+
+    def _get_topic_tagged_articles(self, excluded_pks: list[int]) -> list[StatisticalArticlePage]:
+        # Get descendant series PKs to exclude
+        descendant_series_pks = set(
+            ArticleSeriesPage.objects.descendant_of(self.topic_page).values_list("pk", flat=True)
+        )
+
+        # Get all other article series across the CMS that are tagged with this topic of this topic page
+        # (excluding descendants)
+        tagged_series = (
+            ArticleSeriesPage.objects.live()
+            .public()
+            .filter(topics__topic_id=self.topic_page.topic_id)
+            .exclude(pk__in=descendant_series_pks)
+        )
+
+        latest_pks = self._get_latest_articles_by_series(tagged_series)
+
+        # Get the latest articles, excluding those already highlighted
+        return list(
+            StatisticalArticlePage.objects.filter(pk__in=latest_pks)
+            .filter(locale=self.topic_page.locale)  # Ensure articles are in the same locale as the topic page
+            .exclude(pk__in=excluded_pks)
+            .live()
+            .public()
+            .defer_streamfields()
+            .order_by("-release_date")
+        )
+
+    def _get_latest_articles_by_series(self, series_qs: "PageQuerySet") -> list[int]:
+        """Get the PK of the latest article for each series."""
+        # Subquery to find the latest article that is a descendant of each series
+        newest_qs = (
+            StatisticalArticlePage.objects.live()
+            .public()
+            .filter(path__startswith=OuterRef("path"), depth__gte=OuterRef("depth"))
+            .order_by("-release_date")
+        )
+
+        return list(
+            series_qs.annotate(latest_child_page=Subquery(newest_qs.values("pk")[:1])).values_list(
+                "latest_child_page", flat=True
+            )
+        )
+
+    def _combine_and_sort_articles(self, *articles_lists: list[StatisticalArticlePage]) -> list[StatisticalArticlePage]:
+        """Combine articles lists and sort them by release date."""
+        combined = []
+        for articles in articles_lists:
+            combined.extend(articles)
+
+        combined.sort(key=lambda page: page.release_date, reverse=True)
+        return combined
+
+
+class MethodologyProcessor:
```

Comment:
It would be nice to get the devs' opinions on if it's worth putting these processor service classes in their own file?

Reply:
https://github.com/ONSdigital/dis-wagtail/pull/366#issuecomment-3349251776